### PR TITLE
Switch decorators from Draper to Rails builtin decorations

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [x] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

This PR does the following things:

- uses [delegate_missing_to](https://blog.bigbinary.com/2017/05/30/rails-5-1-adds-delegate-missing-to.html) that was added in Rails 5.1 [as a simpler decoration mechanism](https://api.rubyonrails.org/classes/Module.html#method-i-delegate_missing_to)
- fixes some bugs in the existing decorators
- adds specs for all decorators methods that were missing them
- says bye bye to Draper 🗡 as we use none of its advanced features and draper has too much magic 👀 

## Added tests?

- [x] yes
- [ ] no, because they aren't needed
- [ ] no, because I need help

## Added to documentation?

- [ ] docs.dev.to
- [ ] readme
- [x] no documentation needed

### CODE REVIEW

File: app/decorators/sponsorship_decorator.rb [hunk 1]
```diff
@@ -1,12 +1,11 @@
 class SponsorshipDecorator < ApplicationDecorator
-  delegate_all
-
-  def level_color_hex
+  def level_background_color
```
Comment (rhymes) [inline app/decorators/sponsorship_decorator.rb]: just gave it a better name

File: app/decorators/user_decorator.rb [hunk 2]
```diff
@@ -44,20 +42,21 @@ def enriched_colors
       }
     else
       {
-        bg: bg_color_hex || assigned_color[:bg],
```
Comment (rhymes) [inline app/decorators/user_decorator.rb]: given the precondition, the `OR` is never reachable

File: app/decorators/user_decorator.rb [hunk 3]
```diff
@@ -44,20 +42,21 @@ def enriched_colors
...
 
   def config_body_class
-    body_class = ""
-    body_class += config_theme.tr("_", "-")
-    body_class += " #{config_font.tr('_', '-')}-article-body"
-    body_class += " pro-status-#{pro?}"
-    body_class += " trusted-status-#{trusted}"
-    body_class += " #{config_navbar.tr('_', '-')}-navbar-config"
-    body_class
+    body_class = [
```
Comment (rhymes) [inline app/decorators/user_decorator.rb]: joining lists is faster than building and discarding strings

File: app/models/application_record.rb [hunk 4]
```diff
@@ -20,4 +20,31 @@ def self.estimated_count
...
+  def self.decorator_class(called_on = self)
+    prefix = respond_to?(:model_name) ? model_name : name
+    decorator_name = "#{prefix}Decorator"
+    decorator_name_constant = decorator_name.safe_constantize
+    return decorator_name_constant unless decorator_name_constant.nil?
+
+    return superclass.decorator_class(called_on) if superclass.respond_to?(:decorator_class)
+
+    raise UninferrableDecoratorError, "Could not infer a decorator for #{called_on.class.name}."
+  end
```
Comment (rhymes) [inline app/models/application_record.rb]: these are support methods for "decoratable" objects, as we use decorators only on models, I didn't put it in a separate module, let me know if you'd like me to move that out, Draper has a `Decoratable` module

File: app/models/podcast_episode.rb [hunk 5]
```diff
@@ -98,10 +98,6 @@ def body_text
     ActionView::Base.full_sanitizer.sanitize(processed_html)
   end
 
-  def published_at_date_slashes
```
Comment (rhymes) [inline app/models/podcast_episode.rb]: never used anywhere

File: app/decorators/application_decorator.rb [hunk 6]
```diff
@@ -1,8 +1,17 @@
...
+
+  def initialize(object)
+    @object = object
+  end
+
+  def decorated?
+    true
+  end
+
+  def self.decorate_collection(objects)
```
Comment (joshpuetz) [inline app/decorators/application_decorator.rb]: Thanks for implementing this: I use it in a few places in `StoriesController`!

File: app/models/application_record.rb (hunk 4 above)
Comment (joshpuetz) [reply app/models/application_record.rb]: I don't think this is too bad if this is the only decorator support code that makes it into `ApplicationRecord`. If more methods appear in the future, it could be refactored then!

File: app/controllers/articles_controller.rb [hunk 7]
```diff
@@ -88,14 +88,16 @@ def new
...
     @version = @article.has_frontmatter? ? "v1" : "v2"
     @user = @article.user
     @organizations = @user&.organizations
   end
 
   def manage
-    @article = @article.decorate
     authorize @article
+
+    @article = @article.decorate
```
Comment (juliannatetreault) [inline app/controllers/articles_controller.rb]: Why did you move this line below line 98 rather than leaving it above it on line 97?

File: app/decorators/article_decorator.rb [hunk 8]
```diff
@@ -26,26 +24,29 @@ def url
...
-    elsif article.title.size > 22
+    elsif title.size > 22
       "medium"
     else
       "short"
     end
   end
 
   def internal_utm_params(place = "additional_box")
+    org_slug = organization&.slug
```
Comment (juliannatetreault) [inline app/decorators/article_decorator.rb]: Good call!

File: app/controllers/articles_controller.rb (hunk 7 above)
Comment (rhymes) [reply app/controllers/articles_controller.rb]: Because it broke specs :) We passed a decorated object to the authorization layer, though decoration is mostly for controllers and views. I'm not sure what Draper magic was keeping it working before but instead of going down a rabbit hole of fixing it, I just thought "hey, why do we need the decorated object for auth? shouldn't this be just the record". Once I changed that, the spec passed again :D

File: app/models/podcast_episode.rb (hunk 5 above)
Comment (juliannatetreault) [reply app/models/podcast_episode.rb]: Good call!

File: app/decorators/article_decorator.rb (hunk 8 above)
Comment (vaidehijoshi) [reply app/decorators/article_decorator.rb]: Is it possible for `organization&.slug` to be `nil`? Looking at this method, seems like we could be using a `nil` value for `org_slug` if `boosted_additional_articles` is truthy, right? Which means we could end up with `" _boosted"`.

I think this is how the current code also behaves, but just wondering if that's really the right thing to do here.

File: app/decorators/notification_decorator.rb [hunk 9]
```diff
@@ -1,21 +1,30 @@
...
+    return NOTIFIABLE_STUB.new("", nil) if json_data.blank?
+
+    NOTIFIABLE_STUB.new(json_data[type]["class"]["name"], json_data[type]["id"])
   end
 
+  # returns the type of a milestone notification action,
+  # eg. "Milestone::Reaction::64"
   def milestone_type
-    action.split("::")[1]
+    return "" if action.blank?
```
[the remark this answers is not in the document]
Comment (vaidehijoshi) [reply app/decorators/notification_decorator.rb]: why an empty string and not `nil`?

File: app/decorators/notification_decorator.rb [hunk 10]
```diff
@@ -1,21 +1,30 @@
...
+
+    action.split("::").second
   end
 
+  # returns the count of a milestone notification action,
+  # eg. "Milestone::Reaction::64"
   def milestone_count
-    action.split("::")[2]
+    return "" if action.blank?
```
Comment (vaidehijoshi) [inline app/decorators/notification_decorator.rb]: same question as above :)

File: spec/decorators/notification_decorator_spec.rb [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+    it "returns class name and id for the reactable in a struct" do
+      notification = Notification.send_new_comment_notifications_without_delay(comment)
+
+      result = notification.decorate.mocked_object("user")
+      expect(result.name).to eq("User")
+      expect(result.id).to eq(comment.user.id)
+    end
+
+    it "returns struct class and its name" do
+      notification = Notification.send_new_comment_notifications_without_delay(comment)
```
Comment (vaidehijoshi) [inline spec/decorators/notification_decorator_spec.rb]: might be nice to pull this out into a `let` since we're using it in a few places :) but just a nice to have, not a requirement.

File: spec/services/articles/creator_spec.rb [hunk 12]
```diff
@@ -12,9 +12,10 @@
       end.to change(Article, :count).by(1)
     end
 
-    it "returns an article" do
+    it "returns a non decorated, persisted, article" do
```
Comment (vaidehijoshi) [inline spec/services/articles/creator_spec.rb]:
```suggestion
    it "returns a non decorated, persisted article" do
```

File: spec/services/articles/creator_spec.rb [hunk 13]
```diff
@@ -60,9 +61,10 @@
       end.not_to change(Article, :count)
     end
 
-    it "returns an unsaved article" do
+    it "returns a non decorated, non persisted, article" do
```
Comment (vaidehijoshi) [inline spec/services/articles/creator_spec.rb]:
```suggestion
    it "returns a non decorated, non persisted article" do
```

File: app/decorators/notification_decorator.rb (hunk 9 above)
[the remark this answers is not in the document]
Comment (rhymes) [reply app/decorators/notification_decorator.rb]: as the method returns a string I thought it should stick to strings, but I don't have a strong opinion about this

File: spec/decorators/user_decorator_spec.rb [hunk 14]
```diff
@@ -1,48 +1,92 @@
 require "rails_helper"
 
 RSpec.describe UserDecorator, type: :decorator do
-  let(:user) { build_stubbed(:user) }
+  let_it_be_changeable(:saved_user) { create(:user) }
```
Comment (juliannatetreault) [inline spec/decorators/user_decorator_spec.rb]: I've never seen `let_it_be_changeable` used before--what does it do?

File: app/decorators/article_decorator.rb (hunk 8 above)
Comment (rhymes) [reply app/decorators/article_decorator.rb]: I agree with you but I tried as much as possible not to change anything in the current behavior as this PR should be seamsless. Given we're talking about UTM params which are probably indexed by the Google bot, I also thought it wise not to change that, I'm unsure of the ramifications of changing a string there :)

File: spec/decorators/notification_decorator_spec.rb (hunk 11 above)
Comment (rhymes) [reply spec/decorators/notification_decorator_spec.rb]: good call, i'll write it down for a following PR!

File: spec/services/articles/creator_spec.rb (hunk 12 above)
Comment (rhymes) [reply spec/services/articles/creator_spec.rb]: as you've already approved and this is not a blocker I'm going to wait merging this, as it'll invalidate all the reviews. I'll write it down for a following PR

File: spec/decorators/user_decorator_spec.rb (hunk 14 above)
Comment (rhymes) [reply spec/decorators/user_decorator_spec.rb]: Reloads the object for each test, the contributor that sent the PR that added it wrote a beautiful explanation for the reason why: https://github.com/thepracticaldev/dev.to/pull/5407

You can find the code itself in `spec/support/initializers/test_prof.rb`

File: spec/services/articles/creator_spec.rb (hunk 12 above)
Comment (vaidehijoshi) [reply spec/services/articles/creator_spec.rb]: Sounds good! Definitely don't need to invalidate reviews for this

File: spec/decorators/user_decorator_spec.rb (hunk 14 above)
Comment (juliannatetreault) [reply spec/decorators/user_decorator_spec.rb]: Awesome; thank you!

File: app/models/application_record.rb (hunk 4 above)
Comment (citizen428) [reply app/models/application_record.rb]: I never was a big fan of `.decorate`, as it's really hard to see where each decorator is used without looking at context. At a previous company we had a policy of always explicitly wrapping. We also didn't want to tie decoration to models, they usually already do too much in Rails and we had service objects that returned POROs, which depending on context also got decorated. 

But I understand that that's outside the scope of this PR, so LGTM. Modules that only get included in a single place feel like "cleaning up" by brushing the mess off the table into the drawer anyway 😉

File: spec/decorators/application_decorator_spec.rb [hunk 15]
```diff
@@ -0,0 +1,40 @@
+require "rails_helper"
+
+RSpec.describe ApplicationDecorator, type: :decorator do
+  describe "#object" do
+    it "exposes the decorated object" do
+      obj = Class.new
```
Comment (citizen428) [inline spec/decorators/application_decorator_spec.rb]: This is a bit nitpicky, but when I saw `Class.new`, I wondered if it's significant that it's a class instance and had to read the code to make sure it wasn't. `Object.new` would avoid this confusion.

File: app/decorators/application_decorator.rb [hunk 16]
```diff
@@ -1,8 +1,13 @@
-class ApplicationDecorator < Draper::Decorator
-  # Define methods for all decorated objects.
-  # Helpers are accessed through `helpers` (aka `h`). For example:
-  #
-  #   def percent_amount
-  #     h.number_to_percentage object.amount, precision: 2
-  #   end
+class ApplicationDecorator
+  delegate_missing_to :@object
```
[the remark this answers is not in the document]
Comment (rhymes) [reply app/decorators/application_decorator.rb]: [This article](https://blog.bigbinary.com/2017/05/30/rails-5-1-adds-delegate-missing-to.html) has the following explanation:

> Why not SimpleDelegator
> We at BigBinary have used SimpleDelegator. However one issue with this is that statically we do not know to what object the calls are getting delegated to since at run time the delegator could be anything.

> Why not delegate method
> Delegate method works. However here we need to white list all the methods and in some cases the list can get really long

Here you can also find DHH comments about `SimpleDelegator`: https://github.com/rails/rails/issues/23824#issuecomment-221171899

File: spec/decorators/application_decorator_spec.rb (hunk 15 above)
Comment (rhymes) [reply spec/decorators/application_decorator_spec.rb]: adding this to the list of things to change in a following PR :D

File: app/decorators/application_decorator.rb (hunk 16 above)
[the remark this answers is not in the document]
Comment (citizen428) [reply app/decorators/application_decorator.rb]: I always preferred the explicit whitelisting.  Makes it much easier to make your decorated objects "dumb"  and avoid things like SQL queries triggered from views etc. Anyway, that ship has sailed for Rails, thanks for the context.

File: spec/decorators/application_decorator_spec.rb (hunk 15 above)
Comment (citizen428) [reply spec/decorators/application_decorator_spec.rb]: Sure, no rush.
